[FEATURE] QML chart and drawings widget

signedav opened the pull request:
New type of widget to show graphically appealing and interactive items like charts or technical drawings on the attribute form.

The configuration is reduced to the basics. Including example templates and adding expressions. This could be improved with wizards etc.

Looks like this:
![qml2](https://user-images.githubusercontent.com/28384354/45079754-9cdbeb80-b0f3-11e8-8dba-3e02f0351517.gif)

m-kuhn committed:
Work in progress

m-kuhn committed:
Make string translatable

m-kuhn committed:
Initial work on QML widget configuration

m-kuhn committed:
f

m-kuhn committed:
Add some configuration

m-kuhn committed:
Modernize

m-kuhn committed:
Yay, more QML!

signedav committed:
Configuration and display

works in basic - no attributs. no handy configurator.

signedav committed:
get attributes from current feature

by feature.attribute("outlet_1")

signedav committed:
basics with configurable code and attributes

signedav committed:
open on drop

signedav committed:
preview

signedav committed:
layout and expressions

signedav committed:
expressions

signedav committed:
expressions on qml widget

signedav committed:
improved gui handling

signedav committed:
layout

signedav committed:
docstrings and beautification

signedav committed:
remove unused qmlwidget class

because using qquickwidget directly

signedav committed:
fixed qml example

signedav committed:
return if no widget

signedav committed:
preview feature

m-kuhn commented on src/gui/editorwidgets/qgsqmlwidgetwrapper.h:
```diff
@@ -0,0 +1,69 @@
+};
+
+
+class GUI_EXPORT QmlExpression : public QObject
```
can this be put into the cpp file?
I would keep this as an implementation detail (and not public api)

m-kuhn commented on src/core/qgseditformconfig.cpp:
```diff
@@ -558,13 +558,13 @@ QgsAttributeEditorElement *QgsEditFormConfig::attributeEditorElementFromDomEleme
     newElement = new QgsAttributeEditorField( name, idx, parent );
   }
-  else if ( elem.tagName() == QLatin1String( "attributeEditorRelation" ) )
+  else if ( elem.tagName() == QStringLiteral( "attributeEditorRelation" ) )
```
I think QLatin1String is still recommended for comparisons like here

signedav committed:
docs and replacement of qmlexpression

signedav committed:
moved QmlExpression back to header

signedav committed:
some more fixes

Removed QmlExpression from api
On activation of template box, we overwrite the text in the GUI
The preview is loaded including expression values on start
Takes expression instead of currentText from ExpressionWidget - means it delivers fields in quotes

signedav committed:
spelling and missing case in qgsquick

signedav committed:
spelling american

initialization not initialisation

signedav committed:
doc

signedav committed:
initialization

signedav committed:
add doxygen commands in cpp

wonder-sk commented:
Shall we be also requiring Qt Charts library in the CMakeLists if the sample QML code makes use of it?

wonder-sk commented on src/gui/CMakeLists.txt:
```diff
@@ -733,6 +733,17 @@ SET(QGIS_GUI_MOC_HDRS
   processing/qgsprocessingwidgetwrapperimpl.h
 )
+FIND_PACKAGE(Qt5Qml REQUIRED)
+FIND_PACKAGE(Qt5QuickWidgets REQUIRED)
```
I would put this into root CMakeLists where all other libs are searched for

m-kuhn replied:
This would also make this a requirement for server, no?

wonder-sk replied:
Not if it is in conditioned by WITH_GUI (or WITH_DESKTOP)

m-kuhn replied:
I'm rather -1 on this anyway, keeping things modularized helps to keep on overview of what's used where for future changes.

wonder-sk replied:
I think at least for consistency it makes sense to have it in root cmakelists... and it does not require people to look through N cmakelists to figure out the dependencies...

m-kuhn replied:
On the other hand, there is no need to keep conditionals (if with_gui or with_quick or with_whatevereelse) up to date if it's defined where it's actually required.
I don't think CMakeLists is the place where people (should) look for dependencies. There's the installation instruction, package sources or cmake output.

3nids replied:
There is already a WITH_GUI
https://github.com/qgis/QGIS/blob/master/CMakeLists.txt#L363-L365

m-kuhn replied:
I'm not sure "keeping consistency" is really an argument here.

```
./src/auth/oauth2/CMakeLists.txt:  FIND_PACKAGE(QtKeychain REQUIRED)
./src/auth/oauth2/CMakeLists.txt:  FIND_PACKAGE(O2 REQUIRED)
./src/plugins/globe/CMakeLists.txt:FIND_PACKAGE(OSG REQUIRED)
./src/plugins/globe/CMakeLists.txt:FIND_PACKAGE(OSGEARTH REQUIRED)
./src/plugins/globe/CMakeLists.txt:FIND_PACKAGE(Qt5OpenGL REQUIRED)
./src/plugins/georeferencer/CMakeLists.txt:FIND_PACKAGE(GSL REQUIRED)      # Georeferencer
./src/native/CMakeLists.txt:FIND_PACKAGE(Qt5DBus REQUIRED)
./src/native/CMakeLists.txt:  FIND_PACKAGE(Qt5MacExtras)
./src/native/CMakeLists.txt:  FIND_PACKAGE(Qt5WinExtras)
./src/app/CMakeLists.txt:  FIND_PACKAGE(QwtPolar)
./src/app/CMakeLists.txt:      FIND_PROGRAM(WINDRES windres)
./src/server/CMakeLists.txt:FIND_PACKAGE(Fcgi REQUIRED)
./src/customwidgets/CMakeLists.txt:FIND_PACKAGE(Qt5UiPlugin REQUIRED)
./src/customwidgets/CMakeLists.txt:FIND_PACKAGE(Qt5Designer REQUIRED)
./src/providers/oracle/ocispatial/CMakeLists.txt:FIND_PACKAGE(OCI REQUIRED)
./src/providers/wms/CMakeLists.txt:FIND_PACKAGE(Qt5XmlPatterns REQUIRED)
./src/providers/mdal/CMakeLists.txt:FIND_PACKAGE(HDF5)
./src/providers/mdal/CMakeLists.txt:FIND_PACKAGE(NetCDF)
```

wonder-sk commented:
I have forgotten to add: woohoo! :heart: 

Looking forward to see map canvas from QGIS Quick embedded in attribute forms :-)

If we decide that Qt Charts is a new dependency for QGIS 3.4, finally we will have the primary library of choice for all graphing needs, rather than having a mix of matplotlib, pyqtgraph, plotly, d3, ...

nyalldawson commented on src/gui/editorwidgets/qgsqmlwidgetwrapper.h:
```diff
@@ -0,0 +1,92 @@
+
+  public slots:
+
+    //! passes the \a feature into the context property of the widget
```
No dox for overridden methods

nyalldawson commented:
Just did a quick code standards/format review, looks good to me. Can't comment on the QML/editor widget integration specifics though.

nyalldawson commented:
Just looking at that screencast - do we need some setting for editor widgets "hide on new feature form"? It seems odd to show the empty pie chart when the new point feature is being created, when the chart really only makes sense to show on an existing feature.

signedav commented:
I like the idea with "hide on new feature form". At least until there will probably be in future kind of a "live-update" of QML while editing the feature.

What is @m-kuhn s opinion?

And should it hide the widget (including title) or should the title still be there to see that something is hidden?

nirvn commented:
Hiding widgets on new feature would be good; making widgets dependent (and automatically refresh) on fields would be nice too.

m-kuhn commented:
We could inject a `@form_mode` context variable into expressions (See QgsAttributeForm::Mode => SingleEdit, AddFeature, MultiEdit, Search, AggregateSearch, Identify) and then conditional visibility could be used on any container (and in a followup in the future potentially also directly on widgets)

signedav commented:
We even could use the `@form_mode` in the QML code to set visibility of the item itself to have an "empty" widget on adding feature. 
I support this solution. Only concern is, that it's less easy for a user to understand than a checkbox "hide on new feature". But both is fine for me.

signedav committed:
no case default and removed unused doc

signedav committed:
form_mode for qmlwidgetwrapper expression

signedav committed:
implementation for containers

and changed to string of formMode in the attributeformcontext

signedav committed:
update container visibilty on mode change

m-kuhn commented on src/gui/editorwidgets/qgsqmlwidgetwrapper.h:
```diff
@@ -55,12 +55,14 @@ class GUI_EXPORT QgsQmlWidgetWrapper : public QgsWidgetWrapper
   public slots:
 
-    //! passes the \a feature into the context property of the widget
+    void setQmlContext();
```
private?

m-kuhn commented on src/gui/qgsattributeeditorcontext.h:
```diff
@@ -38,6 +38,18 @@ class GUI_EXPORT QgsAttributeEditorContext
   public:
 
+    //! Form modes
+    enum Mode
```
This is copied from somewhere, correct? Would it be possible to replace the other one with this one in the long run and deprecate it for QGIS 4?

signedav replied:
This should be removed again. I forgot after changing back saving the Mode as `QString` here. Another easy to change possibility would be:
- keep this enum here, work here with enum instead of string for mAttributeFormMode
- use in the `QgsAttributeForm` the enums from here
- implement function `modeString` function here instead of `QgsAttributeForm` and remove there.

What should be prefered?

signedav commented on python/gui/auto_generated/editorwidgets/core/qgswidgetwrapper.sip.in:
```diff
@@ -165,6 +165,15 @@ that any pending changes should be pushed to the edit buffer or they
+
+    void contextChanged();
+%Docstring
+Signal when QgsAttributeEditorContext mContext changed after the widget wrapper has already been intialized
```
Not sure if I understand your question. Ìt can be connected on creation of widget wrappers, in case they need to make an action. Like `QgsQmlWidgetWrapper`...

m-kuhn replied:
I was mostly wondering why the part "after the widget wrapper has already been initialized" is in the doc string.

m-kuhn replied:
So it's mostly a historical consideration?
In this case I don't think it's useful for future readers of this docs.

signedav replied:
No not historical. E.g. the `QgsQmlWidgetWrapper` used the context to set values. On change of the context, these values has to be changed as well. That's why the signal is implemented, to trigger these changes, when the context changes *after* the Widget is initialized. If it changes before, then the inexistent Widged doesn't mind anyway. But probably it's too clear, so it only confuses...

signedav committed:
Enum Mode in QgsAttributeEditorContext

instead of QgsAttributeForm - because it's used on places not including QgsAttributeForm
now it's able to use @form_mode in expressions of containers and QML widgets

m-kuhn commented on src/gui/qgsattributeeditorcontext.h:
```diff
@@ -204,6 +220,27 @@ class GUI_EXPORT QgsAttributeEditorContext
+     * Returns given \a attributeFormMode as string
+     * \since QGIS 3.4
+     */
+    QString attributeFormModeString( const Mode &attributeFormMode )
```
I think this should either be static, or (preferred) no parameter and work on `mAttributeFormMode` instead.

m-kuhn replied:
Oh, and const

m-kuhn commented on src/gui/qgsattributeform.h:
```diff
@@ -46,18 +46,6 @@ class GUI_EXPORT QgsAttributeForm : public QWidget
 
   public:
 
-    //! Form modes
```
To avoid an API break, this needs to stay around until QGIS 4 (but add `\deprecated` to the docs)

signedav committed:
last fixes and sip_include

m-kuhn commented on src/app/qgsattributesformproperties.cpp:
```diff
@@ -956,119 +1006,281 @@ void DnDTree::onItemDoubleClicked( QTreeWidgetItem *item, int column )
+                                      "}\n" ) );
+            break;
+          }
+          default:
```
No enum in place here as far as I can see...

m-kuhn reviewed (approved):
Any objections to hitting the merge button?

wonder-sk commented:
No objection, just a pending question regarding QtCharts library - is it now a required dependency? If it is, we should declare it in CMakeLists too, right?

m-kuhn commented:
IMO it should, I think QtCharts got only added in 5.7, so do we need a `WITH_QTCHARTS` directive?

wonder-sk commented:
I think you are right - before Qt 5.7 the Qt Charts library was versioned and delivered separately:
https://blog.qt.io/blog/2016/01/18/qt-charts-2-1-0-release/

Looks like we still have Qt 5.4 as the minimum version, so if we stick to it, QtCharts should be optional for < 5.7. But given that 5.7 is already more than two years old and even Debian stable ships it, maybe it would be better to raise minimal Qt version to 5.7 for QGIS 3.4?

m-kuhn commented:
I think there were some discussions about the minimum version recently, but I cannot recall the details right now. Let's just add it with a conditional version check (if qt version >= 5.7 require qtcharts) and lead the discussion separately.

signedav committed:
QtChart in CMake file

signedav committed:
include Qt5chars INCLUDE_DIRECTORIES

signedav committed:
add qt5chartslib

m-kuhn commented:
Have a look here `.docker/qgis3-build-deps.dockerfile`

signedav committed:
qt5charts lib in dockerfile

signedav committed:
qt5charts5 library

signedav committed:
more includes

signedav committed:
removed qt5charts stuff from cmakelist

because it's not needed there. should be used in runtime.
so it included qt5charts in installs

signedav commented:
I removed charts stuff from cmakelist because it's not used for build but during run time instead. So added to the package control files.

m-kuhn commented:
And off we go!
